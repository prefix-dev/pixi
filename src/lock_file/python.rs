use crate::project::manifest::SystemRequirements;
use crate::{
    consts::PROJECT_MANIFEST, lock_file::python_name_mapping,
    project::manifest::LibCSystemRequirement, virtual_packages::default_glibc_version, Project,
};
use itertools::Itertools;
use pep508_rs::{MarkerEnvironment, StringVersion};
use rattler_conda_types::{PackageRecord, Platform, RepoDataRecord, Version, VersionWithSource};
use rip::{
    tags::{WheelTag, WheelTags},
    PinnedPackage, SDistResolution,
};
use std::collections::HashMap;
use std::{str::FromStr, vec};

/// Resolve python packages for the specified project.
pub async fn resolve_pypi_dependencies<'p>(
    project: &'p Project,
    platform: Platform,
    conda_packages: &[RepoDataRecord],
) -> miette::Result<Vec<PinnedPackage<'p>>> {
    let dependencies = match project.pypi_dependencies() {
        Some(deps) if !deps.is_empty() => deps,
        _ => return Ok(vec![]),
    };

    let requirements = dependencies
        .iter()
        .map(|(name, req)| req.as_pep508(name))
        .collect::<Vec<pep508_rs::Requirement>>();

    // Determine the python packages that are installed by the conda packages
    let conda_python_packages =
        python_name_mapping::find_conda_python_packages(conda_packages).await?;

    if !conda_python_packages.is_empty() {
        tracing::info!(
            "the following python packages are assumed to be installed by conda: {conda_python_packages}",
            conda_python_packages =
                conda_python_packages
                    .iter()
                    .format_with(", ", |p, f| f(&format_args!(
                        "{name} {version}",
                        name = &p.name,
                        version = &p.version
                    )))
        );
    } else {
        tracing::info!("there are no python packages installed by conda");
    }

    // Determine the python interpreter that is installed as part of the conda packages.
    let python_record = conda_packages
        .iter()
        .find(|r| is_python(r))
        .ok_or_else(|| miette::miette!("could not resolve pypi dependencies because no python interpreter is added to the dependencies of the project.\nMake sure to add a python interpreter to the [dependencies] section of the {PROJECT_MANIFEST}, or run:\n\n\tpixi add python"))?;

    // Determine the environment markers
    let marker_environment = determine_marker_environment(platform, python_record.as_ref())?;

    // Determine the compatible tags
    let compatible_tags = project_platform_tags(
        platform,
        &project.manifest.system_requirements,
        python_record.as_ref(),
    );

    // Resolve the PyPi dependencies
    let mut result = rip::resolve(
        project.pypi_package_db()?,
        &requirements,
        &marker_environment,
        Some(&compatible_tags),
        conda_python_packages
            .into_iter()
            .map(|p| (p.name.clone(), p))
            .collect(),
        HashMap::default(),
        &rip::ResolveOptions {
            // TODO: Change this once we fully support sdists.
            sdist_resolution: SDistResolution::OnlyWheels,
        },
    )
    .await?;

    // Remove any conda package from the result
    result.retain(|p| !p.artifacts.is_empty());

    Ok(result)
}

/// Returns true if the specified record refers to a version/variant of python.
/// TODO: Add support for more variants.
pub fn is_python(record: &RepoDataRecord) -> bool {
    record.package_record.name.as_normalized() == "python"
}

/// Determine the available env markers based on the platform and python package.
pub fn determine_marker_environment(
    platform: Platform,
    python_record: &PackageRecord,
) -> miette::Result<MarkerEnvironment> {
    // Determine system specific information
    let (sys_platform, platform_system, os_name) = if platform.is_linux() {
        ("linux", "Linux", "posix")
    } else if platform.is_osx() {
        ("darwin", "Darwin", "posix")
    } else if platform.is_windows() {
        ("win32", "Windows", "nt")
    } else {
        miette::bail!("could not determine python environment markers for {platform}")
    };

    // Determine implementation name
    let (implementation_name, platform_python_implementation) =
        if python_record.name.as_normalized() == "python" {
            ("cpython", "CPython")
        } else {
            miette::bail!(
                "unsupported python variant {}",
                python_record.name.as_source()
            )
        };

    Ok(MarkerEnvironment {
        implementation_name: String::from(implementation_name),
        implementation_version: version_to_string_version(&python_record.version),
        os_name: String::from(os_name),
        platform_python_implementation: String::from(platform_python_implementation),
        platform_system: String::from(platform_system),
        python_full_version: version_to_string_version(&python_record.version),
        python_version: python_record
            .version
            .version()
            .as_major_minor()
            .and_then(|(major, minor)| StringVersion::from_str(&format!("{major}.{minor}")).ok())
            .ok_or_else(|| {
                miette::miette!(
                    "could not convert python version {}, to a major minor version",
                    &python_record.version
                )
            })?,
        sys_platform: String::from(sys_platform),

        // TODO: Can we figure this out?
        platform_machine: "".to_string(),

        // I assume we can leave these empty
        platform_release: "".to_string(),
        platform_version: "".to_string(),
    })
}

/// Convert a [`VersionWithSource`] to a [`StringVersion`].
fn version_to_string_version(version: &VersionWithSource) -> StringVersion {
    StringVersion::from_str(&version.to_string()).expect("could not convert between versions")
}

/// Returns the compatible tags for the project on the given platform with the given python package.
fn project_platform_tags(
    platform: Platform,
    system_requirements: &SystemRequirements,
    python_record: &PackageRecord,
) -> WheelTags {
    let platforms = project_platforms(platform, system_requirements);

    let mut tags = Vec::new();

    if python_record.name.as_normalized() == "python" {
        tags.append(&mut cpython_tags(&python_record.version, &platforms));
    } else {
        todo!("no support for generic tags yet");
    }

    tags.append(&mut compatible_tags(&python_record.version, &platforms).collect());

    WheelTags::from_iter(tags.into_iter())
}

fn project_platforms(platform: Platform, system_requirements: &SystemRequirements) -> Vec<String> {
    if platform.is_windows() {
        match platform {
            Platform::Win32 => vec![String::from("win32")],
            Platform::Win64 => vec![String::from("win_amd64")],
            Platform::WinArm64 => vec![String::from("win_arm64")],
            _ => unreachable!("not windows"),
        }
    } else if platform.is_linux() {
        let max_glibc_version = match &system_requirements.libc {
            Some(LibCSystemRequirement::GlibC(v)) => v.clone(),
            Some(LibCSystemRequirement::OtherFamily(_)) => return Vec::new(),
            None => default_glibc_version(),
        };
        linux_platform_tags(platform, &max_glibc_version)
    } else {
        todo!("no implementation for mac yet!")
    }
}

/// Returns the platform tags for linux based OS
fn linux_platform_tags(platform: Platform, max_glibc_version: &Version) -> Vec<String> {
    let arch = match platform {
        Platform::Linux32 => "_i686",
        Platform::Linux64 => "_x86_64",
        Platform::LinuxAarch64 => "_aarch64",
        Platform::LinuxArmV7l => "_armv7l",
        Platform::LinuxPpc64le => "_ppc64le",
        Platform::LinuxPpc64 => "_ppc64",
        Platform::LinuxS390X => "_s390x",
        Platform::LinuxRiscv32 => return Vec::new(),
        Platform::LinuxRiscv64 => return Vec::new(),
        Platform::LinuxArmV6l => return Vec::new(),
        _ => unreachable!("not linux"),
    };
    manylinux_versions(max_glibc_version)
        .into_iter()
        .map(|p| format!("{p}{arch}"))
        .collect()
}

// Generate all manylinux tags based on the major minor version of glibc.
fn manylinux_versions(glibc_version: &Version) -> Vec<String> {
    let Some((major, minor)) = glibc_version.as_major_minor() else {
        return Vec::new();
    };

    let mut result = Vec::new();
    for minor in (0..=minor).rev() {
        result.push(format!("manylinux_{major}_{minor}"));
        match (major, minor) {
            (2, 5) => result.push(String::from("manylinux1")),
            (2, 12) => result.push(String::from("manylinux2010")),
            (2, 17) => result.push(String::from("manylinux2014")),
            _ => {}
        }
    }

    result
}

/// Returns all the tags a specific cpython implementation supports.
fn cpython_tags<P, PIter>(python_version: &Version, platforms: PIter) -> Vec<WheelTag>
where
    P: Into<String>,
    PIter: IntoIterator<Item = P>,
    PIter::IntoIter: Clone,
{
    let Some((major, minor)) = python_version.as_major_minor() else {
        return Vec::new();
    };

    let interpreter = format!("cp{major}{minor}");
    let core_abi = format!("cp{}{}", major, minor);

    let mut result = Vec::new();
    let platforms = platforms.into_iter().map(Into::into);

    // Add the core tags
    result.extend(platforms.clone().map(|platform| WheelTag {
        interpreter: interpreter.clone(),
        abi: core_abi.clone(),
        platform,
    }));

    // Add the "abi3"  tags
    if python_abi3_applies(major, minor) {
        result.extend(platforms.clone().map(|platform| WheelTag {
            interpreter: interpreter.clone(),
            abi: String::from("abi3"),
            platform,
        }));
    }

    // Add the "none" abi tags
    result.extend(platforms.clone().map(|platform| WheelTag {
        interpreter: interpreter.clone(),
        abi: String::from("none"),
        platform,
    }));

    // Add other abi3 compatible cpython versions
    if python_abi3_applies(major, minor) {
        for minor in (2..minor).rev() {
            let interpreter = format!("cp{major}{minor}");
            result.extend(platforms.clone().map(|platform| WheelTag {
                interpreter: interpreter.clone(),
                abi: String::from("abi3"),
                platform,
            }));
        }
    }

    result
}

/// Returns true if the specified Python version supports abi3. PEP 384 was first implemented in
/// Python 3.2.
fn python_abi3_applies(major: u64, minor: u64) -> bool {
    major >= 3 && minor >= 2
}

/// Returns an iterator that yields all the different versions of the python interpreter.
fn py_interpreter_range(python_version: &Version) -> impl Iterator<Item = String> {
    let python_version = python_version.as_major_minor();

    let core_version = python_version
        .into_iter()
        .map(|(major, minor)| format!("py{major}{minor}"));
    let major_version = python_version
        .into_iter()
        .map(|(major, _)| format!("py{major}"));
    let minor_versions = python_version.into_iter().flat_map(|(major, minor)| {
        (0..minor)
            .rev()
            .map(move |minor| format!("py{major}{minor}"))
    });
    core_version.chain(major_version).chain(minor_versions)
}

/// Returns compatible tag for any interpreter.
fn compatible_tags<'a, P, PIter>(
    python_version: &'a Version,
    platforms: PIter,
) -> impl Iterator<Item = WheelTag> + 'a
where
    P: Into<String>,
    PIter: IntoIterator<Item = P>,
    PIter::IntoIter: Clone + 'a,
{
    py_interpreter_range(python_version)
        .cartesian_product(platforms.into_iter())
        .map(|(interpreter, platform)| WheelTag {
            interpreter,
            abi: String::from("none"),
            platform: platform.into(),
        })
        .chain(
            py_interpreter_range(python_version).map(|interpreter| WheelTag {
                interpreter,
                abi: String::from("none"),
                platform: String::from("any"),
            }),
        )
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_cpython_tags() {
        let tags: Vec<_> = cpython_tags(&Version::from_str("3.11.2").unwrap(), vec!["win_amd64"])
            .into_iter()
            .map(|t| t.to_string())
            .collect();
        insta::assert_debug_snapshot!(tags);
    }

    #[test]
    fn test_py_interpreter_range() {
        let tags: Vec<_> = py_interpreter_range(&Version::from_str("3.11.2").unwrap()).collect();
        insta::assert_debug_snapshot!(tags);
    }

    #[test]
    fn test_compatible_tags() {
        let tags: Vec<_> =
            compatible_tags(&Version::from_str("3.11.2").unwrap(), vec!["win_amd64"])
                .map(|t| t.to_string())
                .collect();
        insta::assert_debug_snapshot!(tags);
    }

    #[test]
    fn test_linux_platform_tags() {
        let tags: Vec<_> =
            linux_platform_tags(Platform::Linux64, &Version::from_str("2.17").unwrap())
                .into_iter()
                .map(|t| t.to_string())
                .collect();
        insta::assert_debug_snapshot!(tags);
    }
}
