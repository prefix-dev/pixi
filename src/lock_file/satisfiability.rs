use super::{PypiRecord, PypiRecordsByName, RepoDataRecordsByName};
use crate::project::manifest::python::AsPep508Error;
use crate::{project::Environment, pypi_marker_env::determine_marker_environment};
use distribution_types::DirectGitUrl;
use itertools::Itertools;
use miette::Diagnostic;
use pep440_rs::VersionSpecifiers;
use pep508_rs::Requirement;
use rattler_conda_types::ParseStrictness::Lenient;
use rattler_conda_types::{
    GenericVirtualPackage, MatchSpec, ParseMatchSpecError, Platform, RepoDataRecord,
};
use rattler_lock::{ConversionError, Package, PypiPackageData, UrlOrPath};
use std::path::Path;
use std::str::FromStr;
use std::{
    borrow::Cow,
    collections::{HashMap, HashSet},
};
use thiserror::Error;
use url::Url;
use uv_normalize::{ExtraName, PackageName};

#[derive(Debug, Error, Diagnostic)]
pub enum EnvironmentUnsat {
    #[error("the channels in the lock-file do not match the environments channels")]
    ChannelsMismatch,
}

#[derive(Debug, Error, Diagnostic)]
pub enum PlatformUnsat {
    #[error("the requirement '{0}' could not be satisfied (required by '{1}')")]
    UnsatisfiableMatchSpec(MatchSpec, String),

    #[error("the requirement '{0}' could not be satisfied (required by '{1}')")]
    UnsatisfiableRequirement(Requirement, String),

    #[error("there was a duplicate entry for '{0}'")]
    DuplicateEntry(String),

    #[error("the requirement '{0}' failed to parse")]
    FailedToParseMatchSpec(String, #[source] ParseMatchSpecError),

    #[error("there are more conda packages in the lock-file than are used by the environment")]
    TooManyCondaPackages,

    #[error("corrupted lock-file entry for '{0}'")]
    CorruptedEntry(String, ConversionError),

    #[error("there are more pypi packages in the lock-file than are used by the environment: {}", .0.iter().format(", "))]
    TooManyPypiPackages(Vec<PackageName>),

    #[error("there are PyPi dependencies but a python interpreter is missing from the lock-file")]
    MissingPythonInterpreter,

    #[error(
        "a marker environment could not be derived from the python interpreter in the lock-file"
    )]
    FailedToDetermineMarkerEnvironment(#[source] Box<dyn Diagnostic + Send + Sync>),

    #[error("{0} requires python version {1} but the python interpreter in the lock-file has version {2}")]
    PythonVersionMismatch(PackageName, VersionSpecifiers, Box<pep440_rs::Version>),

    #[error("when converting {0} into a pep508 requirement")]
    AsPep508Error(PackageName, #[source] AsPep508Error),
}

/// Verifies that all the requirements of the specified `environment` can be satisfied with the
/// packages present in the lock-file.
///
/// This function returns a [`EnvironmentUnsat`] error if a verification issue occurred. The
/// [`EnvironmentUnsat`] error should contain enough information for the user and developer to
/// figure out what went wrong.
pub fn verify_environment_satisfiability(
    environment: &Environment<'_>,
    locked_environment: &rattler_lock::Environment,
) -> Result<(), EnvironmentUnsat> {
    // Check if the channels in the lock file match our current configuration. Note that the order
    // matters here. If channels are added in a different order, the solver might return a different
    // result.
    let channels = environment
        .channels()
        .into_iter()
        .map(|channel| rattler_lock::Channel::from(channel.base_url().to_string()))
        .collect_vec();
    if !locked_environment.channels().eq(&channels) {
        return Err(EnvironmentUnsat::ChannelsMismatch);
    }

    Ok(())
}

/// Verifies that the package requirements of the specified `environment` can be satisfied with the
/// packages present in the lock-file.
///
/// Both Conda and pypi packages are verified by this function. First all the conda package are
/// verified and then all the pypi packages are verified. This is done so that if we can check if
/// we only need to update the pypi dependencies or also the conda dependencies.
///
/// This function returns a [`PlatformUnsat`] error if a verification issue occurred. The
/// [`PlatformUnsat`] error should contain enough information for the user and developer to figure
/// out what went wrong.
pub fn verify_platform_satisfiability(
    environment: &Environment<'_>,
    locked_environment: &rattler_lock::Environment,
    platform: Platform,
    project_root: &Path,
) -> Result<(), PlatformUnsat> {
    // Convert the lock file into a list of conda and pypi packages
    let mut conda_packages: Vec<RepoDataRecord> = Vec::new();
    let mut pypi_packages: Vec<PypiRecord> = Vec::new();
    for package in locked_environment.packages(platform).into_iter().flatten() {
        match package {
            Package::Conda(conda) => {
                let url = conda.url().clone();
                conda_packages.push(
                    conda
                        .try_into()
                        .map_err(|e| PlatformUnsat::CorruptedEntry(url.to_string(), e))?,
                );
            }
            Package::Pypi(pypi) => {
                pypi_packages.push((pypi.data().package.clone(), pypi.data().environment.clone()));
            }
        }
    }

    // Create a lookup table from package name to package record. Returns an error if we find a
    // duplicate entry for a record
    let repodata_records_by_name = match RepoDataRecordsByName::from_unique_iter(conda_packages) {
        Ok(conda_packages) => conda_packages,
        Err(duplicate) => {
            return Err(PlatformUnsat::DuplicateEntry(
                duplicate.package_record.name.as_source().to_string(),
            ))
        }
    };

    // Create a lookup table from package name to package record. Returns an error if we find a
    // duplicate entry for a record
    let pypi_records_by_name = match PypiRecordsByName::from_unique_iter(pypi_packages) {
        Ok(conda_packages) => conda_packages,
        Err(duplicate) => return Err(PlatformUnsat::DuplicateEntry(duplicate.0.name.to_string())),
    };

    verify_package_platform_satisfiability(
        environment,
        &repodata_records_by_name,
        &pypi_records_by_name,
        platform,
        project_root,
    )
}

enum Dependency {
    Conda(MatchSpec, Cow<'static, str>),
    PyPi(Requirement, Cow<'static, str>),
}

/// Check satatisfiability of a pypi requirement against a locked pypi package
/// This also does an additional check for git urls when using direct url references
pub fn pypi_satifisfies(locked_data: &PypiPackageData, spec: &Requirement) -> bool {
    // Check if the name matches
    if spec.name != locked_data.name {
        return false;
    }

    // Check if the version of the requirement matches
    match &spec.version_or_url {
        None => true,
        Some(pep508_rs::VersionOrUrl::Url(spec_url)) => {
            // In the case that both the spec and the locked data are direct git urls
            // we need to compare the urls to see if they are the same
            let spec_git_url = DirectGitUrl::try_from(&spec_url.to_url()).ok();
            let locked_git_url = locked_data
                .url_or_path
                .as_url()
                .and_then(|url| DirectGitUrl::try_from(url).ok());

            // Both are git url's
            if let (Some(spec_git_url), Some(locked_data_url)) = (spec_git_url, locked_git_url) {
                let base_is_same =
                    spec_git_url.url.repository() == locked_data_url.url.repository();

                // If the spec does not specify a revision than any will do
                // E.g `git.com/user/repo` is the same as `git.com/user/repo@adbdd`
                if spec_git_url.url.reference().is_none() {
                    base_is_same
                } else {
                    // If the spec does specify a revision than the revision must match
                    base_is_same && spec_git_url.url.reference() == locked_data_url.url.reference()
                }
            } else {
                let spec_path_or_url = spec_url
                    .given()
                    .and_then(|url| UrlOrPath::from_str(url).ok())
                    .unwrap_or(UrlOrPath::Url(spec_url.to_url()));

                // Strip the direct+ prefix if it exists for the direct url
                // because this is not part of the `Requirement` spec
                // we use this to record that it is a direct url
                let locked_path_or_url = match locked_data.url_or_path.clone() {
                    UrlOrPath::Url(url) => UrlOrPath::Url(
                        url.as_ref()
                            .strip_prefix("direct+")
                            .and_then(|str| Url::parse(str).ok())
                            .unwrap_or(url),
                    ),
                    UrlOrPath::Path(path) => UrlOrPath::Path(path),
                };
                spec_path_or_url == locked_path_or_url
            }
        }
        Some(pep508_rs::VersionOrUrl::VersionSpecifier(spec)) => {
            spec.contains(&locked_data.version)
        }
    }
}

pub fn verify_package_platform_satisfiability(
    environment: &Environment<'_>,
    locked_conda_packages: &RepoDataRecordsByName,
    locked_pypi_environment: &PypiRecordsByName,
    platform: Platform,
    project_root: &Path,
) -> Result<(), PlatformUnsat> {
    // Determine the dependencies requested by the environment
    let conda_specs = environment
        .dependencies(None, Some(platform))
        .into_match_specs()
        .map(|spec| Dependency::Conda(spec, "<environment>".into()))
        .collect_vec();

    if conda_specs.is_empty() && !locked_conda_packages.is_empty() {
        return Err(PlatformUnsat::TooManyCondaPackages);
    }

    let pypi_requirements = environment
        .pypi_dependencies(Some(platform))
        .iter()
        .flat_map(|(name, reqs)| {
            reqs.iter().map(move |req| {
                Ok::<Dependency, PlatformUnsat>(Dependency::PyPi(
                    req.as_pep508(name.as_normalized(), project_root)
                        .map_err(|e| {
                            PlatformUnsat::AsPep508Error(name.as_normalized().clone(), e)
                        })?,
                    "<environment>".into(),
                ))
            })
        })
        .collect::<Result<Vec<_>, _>>()?;

    if pypi_requirements.is_empty() && !locked_pypi_environment.is_empty() {
        return Err(PlatformUnsat::TooManyPypiPackages(
            locked_pypi_environment.names().cloned().collect(),
        ));
    }

    // Create a list of virtual packages by name
    let virtual_packages = environment
        .virtual_packages(platform)
        .into_iter()
        .map(|vpkg| (vpkg.name.clone(), vpkg))
        .collect::<HashMap<_, _>>();

    // Find the python interpreter from the list of conda packages. Note that this refers to the
    // locked python interpreter, it might not match the specs from the environment. That is ok
    // because we will find that out when we check all the records.
    let python_interpreter_record = locked_conda_packages.python_interpreter_record();

    // Determine the marker environment from the python interpreter package.
    let marker_environment = python_interpreter_record
        .map(|interpreter| determine_marker_environment(platform, &interpreter.package_record))
        .transpose()
        .map_err(|err| PlatformUnsat::FailedToDetermineMarkerEnvironment(err.into()))?;

    // Determine the pypi packages provided by the locked conda packages.
    let locked_conda_pypi_packages = locked_conda_packages.by_pypi_name();

    // Keep a list of all conda packages that we have already visisted
    let mut conda_packages_visited = HashSet::new();
    let mut pypi_packages_visited = HashSet::new();
    let mut pypi_requirements_visited = pypi_requirements
        .iter()
        .filter_map(|r| match r {
            Dependency::PyPi(req, _) => Some(req.clone()),
            _ => None,
        })
        .collect::<HashSet<_>>();

    // Iterate over all packages. First iterate over all conda matchspecs and then over all pypi
    // requirements. We want to ensure we always check the conda packages first.
    let mut conda_queue = conda_specs;
    let mut pypi_queue = pypi_requirements;
    while let Some(package) = conda_queue.pop().or_else(|| pypi_queue.pop()) {
        enum FoundPackage {
            Conda(usize),
            PyPi(usize, Vec<ExtraName>),
        }

        // Determine the package that matches the requirement of matchspec.
        let found_package = match package {
            Dependency::Conda(spec, source) => {
                match &spec.name {
                    None => {
                        // No name means we have to find any package that matches the spec.
                        match locked_conda_packages
                            .records
                            .iter()
                            .position(|record| record.matches(&spec))
                        {
                            None => {
                                // No records match the spec.
                                return Err(PlatformUnsat::UnsatisfiableMatchSpec(
                                    spec,
                                    source.into_owned(),
                                ));
                            }
                            Some(idx) => FoundPackage::Conda(idx),
                        }
                    }
                    Some(name) => {
                        match locked_conda_packages
                            .index_by_name(name)
                            .map(|idx| (idx, &locked_conda_packages.records[idx]))
                        {
                            Some((idx, record)) if record.matches(&spec) => {
                                FoundPackage::Conda(idx)
                            }
                            Some(_) => {
                                // The record does not match the spec, the lock-file is inconsistent.
                                return Err(PlatformUnsat::UnsatisfiableMatchSpec(
                                    spec,
                                    source.into_owned(),
                                ));
                            }
                            None => {
                                // Check if there is a virtual package by that name
                                if let Some(vpkg) = virtual_packages.get(name.as_normalized()) {
                                    if vpkg.matches(&spec) {
                                        // The matchspec matches a virtual package. No need to propagate the dependencies.
                                        continue;
                                    } else {
                                        // The record does not match the spec, the lock-file is inconsistent.
                                        return Err(PlatformUnsat::UnsatisfiableMatchSpec(
                                            spec,
                                            source.into_owned(),
                                        ));
                                    }
                                } else {
                                    // The record does not match the spec, the lock-file is inconsistent.
                                    return Err(PlatformUnsat::UnsatisfiableMatchSpec(
                                        spec,
                                        source.into_owned(),
                                    ));
                                }
                            }
                        }
                    }
                }
            }
            Dependency::PyPi(requirement, source) => {
                // Check if there is a pypi identifier that matches our requirement.
                if let Some((identifier, repodata_idx, _)) =
                    locked_conda_pypi_packages.get(&requirement.name)
                {
                    if identifier.satisfies(&requirement) {
                        FoundPackage::Conda(*repodata_idx)
                    } else {
                        // The record does not match the spec, the lock-file is inconsistent.
                        return Err(PlatformUnsat::UnsatisfiableRequirement(
                            requirement,
                            source.into_owned(),
                        ));
                    }
                } else if let Some(idx) = locked_pypi_environment.index_by_name(&requirement.name) {
                    let record = &locked_pypi_environment.records[idx];
                    if pypi_satifisfies(&record.0, &requirement) {
                        FoundPackage::PyPi(idx, requirement.extras)
                    } else {
                        // The record does not match the spec, the lock-file is inconsistent.
                        return Err(PlatformUnsat::UnsatisfiableRequirement(
                            requirement,
                            source.into_owned(),
                        ));
                    }
                } else {
                    // The record does not match the spec, the lock-file is inconsistent.
                    return Err(PlatformUnsat::UnsatisfiableRequirement(
                        requirement,
                        source.into_owned(),
                    ));
                }
            }
        };

        // Add all the requirements of the package to the queue.
        match found_package {
            FoundPackage::Conda(idx) => {
                if !conda_packages_visited.insert(idx) {
                    // We already visited this package, so we can skip adding its dependencies to the queue
                    continue;
                }

                let record = &locked_conda_packages.records[idx];
                for depends in &record.package_record.depends {
                    let spec = MatchSpec::from_str(depends.as_str(), Lenient)
                        .map_err(|e| PlatformUnsat::FailedToParseMatchSpec(depends.clone(), e))?;
                    conda_queue.push(Dependency::Conda(
                        spec,
                        Cow::Owned(record.file_name.clone()),
                    ));
                }
            }
            FoundPackage::PyPi(idx, extras) => {
                let record = &locked_pypi_environment.records[idx];

                // If there is no marker environment there is no python version
                let Some(marker_environment) = marker_environment.as_ref() else {
                    return Err(PlatformUnsat::MissingPythonInterpreter);
                };

                if pypi_packages_visited.insert(idx) {
                    // Ensure that the record matches the currently selected interpreter.
                    if let Some(python_version) = &record.0.requires_python {
                        if !python_version.contains(&marker_environment.python_full_version.version)
                        {
                            return Err(PlatformUnsat::PythonVersionMismatch(
                                record.0.name.clone(),
                                python_version.clone(),
                                marker_environment
                                    .python_full_version
                                    .version
                                    .clone()
                                    .into(),
                            ));
                        }
                    }
                }

                // Add all the requirements of the package to the queue.
                for requirement in &record.0.requires_dist {
                    if !pypi_requirements_visited.insert(requirement.clone()) {
                        continue;
                    }

                    // Skip this requirement if it does not apply.
                    if !requirement.evaluate_markers(marker_environment, &extras) {
                        continue;
                    }

                    pypi_queue.push(Dependency::PyPi(
                        requirement.clone(),
                        record.0.name.as_ref().to_string().into(),
                    ));
                }
            }
        }
    }

    // Check if all locked packages have also been visisted
    if conda_packages_visited.len() != locked_conda_packages.len() {
        return Err(PlatformUnsat::TooManyCondaPackages);
    }

    if pypi_packages_visited.len() != locked_pypi_environment.len() {
        return Err(PlatformUnsat::TooManyPypiPackages(
            locked_pypi_environment
                .names()
                .enumerate()
                .filter_map(|(idx, name)| {
                    if pypi_packages_visited.contains(&idx) {
                        None
                    } else {
                        Some(name.clone())
                    }
                })
                .collect(),
        ));
    }

    Ok(())
}

trait MatchesMatchspec {
    fn matches(&self, spec: &MatchSpec) -> bool;
}

impl MatchesMatchspec for RepoDataRecord {
    fn matches(&self, spec: &MatchSpec) -> bool {
        if !spec.matches(&self.package_record) {
            return false;
        }

        // TODO: We should really move this into rattler
        // Check the channel
        if let Some(channel) = &spec.channel {
            if !self.url.as_str().starts_with(channel.base_url.as_str()) {
                return false;
            }
        }

        true
    }
}

impl MatchesMatchspec for GenericVirtualPackage {
    fn matches(&self, spec: &MatchSpec) -> bool {
        if let Some(name) = &spec.name {
            if name != &self.name {
                return false;
            }
        }

        if let Some(version) = &spec.version {
            if !version.matches(&self.version) {
                return false;
            }
        }

        if let Some(build) = &spec.build {
            if !build.matches(&self.build_string) {
                return false;
            }
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Project;
    use miette::{IntoDiagnostic, NarratableReportHandler};
    use pep440_rs::Version;
    use rattler_lock::LockFile;
    use rstest::rstest;
    use std::{path::PathBuf, str::FromStr};

    #[derive(Error, Debug, Diagnostic)]
    enum LockfileUnsat {
        #[error("environment '{0}' is missing")]
        EnvironmentMissing(String),

        #[error("environment '{0}' does not satisfy the requirements of the project")]
        Environment(String, #[source] EnvironmentUnsat),

        #[error(
            "environment '{0}' does not satisfy the requirements of the project for platform '{1}"
        )]
        PlatformUnsat(String, Platform, #[source] PlatformUnsat),
    }

    fn verify_lockfile_satisfiability(
        project: &Project,
        lock_file: &LockFile,
    ) -> Result<(), LockfileUnsat> {
        for env in project.environments() {
            let locked_env = lock_file
                .environment(env.name().as_str())
                .ok_or_else(|| LockfileUnsat::EnvironmentMissing(env.name().to_string()))?;
            verify_environment_satisfiability(&env, &locked_env)
                .map_err(|e| LockfileUnsat::Environment(env.name().to_string(), e))?;

            for platform in env.platforms() {
                verify_platform_satisfiability(&env, &locked_env, platform, project.root())
                    .map_err(|e| {
                        LockfileUnsat::PlatformUnsat(env.name().to_string(), platform, e)
                    })?;
            }
        }
        Ok(())
    }

    #[rstest]
    fn test_good_satisfiability(
        #[files("tests/satisfiability/*/pixi.toml")] manifest_path: PathBuf,
    ) {
        let project = Project::load(&manifest_path).unwrap();
        let lock_file = LockFile::from_path(&project.lock_file_path()).unwrap();
        match verify_lockfile_satisfiability(&project, &lock_file).into_diagnostic() {
            Ok(()) => {}
            Err(e) => panic!("{e:?}"),
        }
    }

    #[rstest]
    fn test_example_satisfiability(#[files("examples/*/pixi.toml")] manifest_path: PathBuf) {
        let project = Project::load(&manifest_path).unwrap();
        let lock_file = LockFile::from_path(&project.lock_file_path()).unwrap();
        match verify_lockfile_satisfiability(&project, &lock_file).into_diagnostic() {
            Ok(()) => {}
            Err(e) => panic!("{e:?}"),
        }
    }

    #[test]
    fn test_failing_satisiability() {
        let report_handler = NarratableReportHandler::new().with_cause_chain();

        insta::glob!("../../tests/non-satisfiability", "*/pixi.toml", |path| {
            let project = Project::load(path).unwrap();
            let lock_file = LockFile::from_path(&project.lock_file_path()).unwrap();
            let err = verify_lockfile_satisfiability(&project, &lock_file)
                .expect_err("expected failing satisfiability");

            let mut s = String::new();
            report_handler.render_report(&mut s, &err).unwrap();
            insta::assert_snapshot!(s);
        });
    }

    #[test]
    fn test_pypi_git_check_with_rev() {
        // Mock locked datga
        let locked_data = PypiPackageData {
            name: "mypkg".parse().unwrap(),
            version: Version::from_str("0.1.0").unwrap(),
            url_or_path: "git+https://github.com/mypkg@abcd"
                .parse()
                .expect("failed to parse url"),
            hash: None,
            requires_dist: vec![],
            requires_python: None,
        };
        let spec = Requirement::from_str("mypkg @ git+https://github.com/mypkg@abcd").unwrap();
        // This should satisfy:
        assert!(pypi_satifisfies(&locked_data, &spec));
        let non_matching_spec =
            Requirement::from_str("mypkg @ git+https://github.com/mypkg@defgd").unwrap();
        // This should not
        assert!(!pypi_satifisfies(&locked_data, &non_matching_spec));
        // Removing the rev from the Requirement should satisfy any revision
        let spec = Requirement::from_str("mypkg @ git+https://github.com/mypkg").unwrap();
        assert!(pypi_satifisfies(&locked_data, &spec));
    }
}
