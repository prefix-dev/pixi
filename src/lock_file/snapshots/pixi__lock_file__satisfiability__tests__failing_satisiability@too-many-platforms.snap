---
source: src/lock_file/satisfiability.rs
expression: s
---
environment 'default' does not satisfy the requirements of the project
    Diagnostic severity: error
    Caused by: platform(s) 'osx-arm64' present in the lock-file but not in the environment
